macros/src/lib.rs

```diff
@@ -52,35 +73,33 @@ use rust_decimal::Decimal;
 /// assert_eq!("1.2345", number.to_string());
 /// let number = dec!(-5.4321);
 /// assert_eq!("-5.4321", number.to_string());
+/// let number = dec!(-0o1_777);
+/// assert_eq!("-1023", number.to_string());
 /// ```
+///
 #[proc_macro]
 pub fn dec(input: TokenStream) -> TokenStream {
-    let mut source = input.to_string();
+    // Parse the input using our custom parser
+    let dec_input = parse_macro_input!(input as DecInputParser);
+    let value = dec_input.value.as_str();
 
-    // If it starts with `- ` then get rid of the extra space
-    // to_string will put a space between tokens
-    if source.starts_with("- ") {
-        source.remove(1);
-    }
-
-    let decimal = if source.contains('e') || source.contains('E') {
-        match Decimal::from_scientific(&source[..]) {
-            Ok(d) => d,
-            Err(e) => panic!("{}", e),
-        }
+    // Process the parsed input
+    let result = if let Some(radix) = dec_input.radix {
+        str::parse_decimal_with_radix(value, dec_input.exp.unwrap_or_default(), radix)
     } else {
-        match Decimal::from_str_exact(&source[..]) {
-            Ok(d) => d,
-            Err(e) => panic!("{}", e),
-        }
+        str::parse_decimal(value, dec_input.exp.unwrap_or_default())
+    };
+
+    let unpacked = match result {
+        Ok(d) => d,
+        Err(e) => panic!("{}", e),
     };
 
-    let unpacked = decimal.unpack();
     expand(
-        unpacked.lo,
-        unpacked.mid,
-        unpacked.hi,
-        unpacked.negative,
+        unpacked.lo(),
```

Comment:
Unnecessary since the i128 constructor is a const - that said, it doesn't hurt either.